p2panda-sync/src/lib.rs

```diff
@@ -42,21 +42,39 @@ pub trait SyncProtocol<'a>: Send + Sync + Debug {
 
 #[derive(Error, Debug)]
 pub enum SyncError {
-    /// Error which can occur in a running sync session.
-    #[error("sync protocol error: {0}")]
-    Protocol(String),
+    /// Error due to unexpected (buggy or malicious) behaviour of the remote peer.
+    ///
+    /// Indicates that the sync protocol was not correctly followed, for example due to unexpected
+    /// or missing messages, invalid encoding, etc.
+    ///
+    /// Can be used to re-attempt syncing with this peer or down-grading it in priority,
+    /// potentially deny-listing if communication failed too often.
+    #[error("sync session failed due to unexpected protocol behaviour of remote peer: {0}")]
+    RemoteUnexpectedBehaviour(String),
 
-    /// I/O error which occurs during stream handling.
-    #[error("input/output error: {0}")]
-    IoError(#[from] std::io::Error),
+    /// Remote peer didn't show any activity for some time.
+    ///
+    /// Can be used to re-attempt syncing with this peer or down-grading it in priority,
+    /// potentially deny-listing if communication failed too often.
+    #[error("sync session failed due to remote peer timeout")]
+    RemoteTimeout,
 
-    /// Error which occurs when encoding or decoding protocol messages.
-    #[error("codec error: {0}")]
-    Codec(String),
+    /// Critical error due to system failure on our end.
+    ///
+    /// This indicates that our system is running out of resources (storage layer failure etc.) or
+    /// we have a buggy implementation.
+    #[error("sync session failed due critical system error: {0}")]
+    Critical(String),
```

Comment:
Since `RemoteTimeout` is more a connection-layer concern, I'm wondering if we can have that on a higher-level in the connection manager?

Not saying we need this already, but more if it's possible.

Reply:
Thanks for looking into these error types, really like the changes you made so far, definitely more expressive for users + implementers. 

- `RemoteUnexpectedBehaviour`: I wonder if this could be split into two distinct types representing: `1)` the remote sent badly formed data; `2)` the remote sent a valid message but it violated the protocol in some way. `2)` is still very broad (could be unexpected message order, or access-control violation) and it might be that we add useful variants later, but maybe to start that's enough. Naming wise, maybe `RemoteInvalidMessage` & `RemoteUnexpectedBehaviour`?  
- `RemoteTimeout`: I agree this could be moved into a higher layer, currently `-sync` doesn't really know about "connections" exactly... but I also wouldn't mind leaving it here. If it stayed here, maybe `RemoteUnresponsive` could be good?

Reply:
I'm happy with these changes. Regarding the `RemoteTimeout` variant, I can see it living in either place but probably lean towards moving it into `SyncAttemptError` in the sync manager. Then we would just be concerned whether the `connect_and_sync()` attempt completed in the given time limit or not.

Reply:
> Regarding the RemoteTimeout variant, I can see it living in either place but probably lean towards moving it into SyncAttemptError in the sync manager. Then we would just be concerned whether the connect_and_sync() attempt completed in the given time limit or not.

I was wondering that as well. I think the `RemoteTimeout` makes sense if you can actually measure the timing _between_ messages (which I'm not sure the connection manager can see), but I think that's already too much entering "nice to have" territory. I'll leave it out for now and we can reason about timeout handling in another PR, maybe it's better to have on connection handler level.